File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        pair = {')': '(', ']': '[', '}': '{'}

        for c in s:
            # Check if c is valid character
            if c not in "()[]{}":
                raise ValueError(f"Invalid character: {c}")

            if c in pair.values():  # left bracket
                st.append(c)
            else:  # right bracket
                if not st or st[-1] != pair[c]:
                    return False
                st.pop()

        return not st

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
#include <stack>

class Solution {
public:
    bool isValid(string s) {
        // ASCII: 40 ( 41 ) 91 [ 93 ] 123 { 125 }
        stack<char> st;
        for (auto ch : s) {
            // first we peak the stack
            if (!st.empty()) {
                if (st.top() == ch - 1 || st.top() == ch - 2)
                    st.pop();
                else
                    st.push(ch);  
            }     
            else
                st.push(ch);  
        }
        
        if (st.empty())
            return true;
        else
            return false;
    }
};